docs/kubernetes-reliability.md

```diff
@@ -0,0 +1,101 @@
+Distributed system such as Kubernetes are designed to be resilient to the
+failures.  More details about Kubernetes High-Availability (HA) may be found at
+[Building High-Availability Clusters](https://kubernetes.io/docs/admin/high-availability/)
+
+To have a simple view the most of parts of HA will be skipped to describe
+Kubelet<->Controller Manager communication only.
+
+By default the normal behavior looks like:
+
+1. Kubelet updates its status to Master's API Server periodically described at
+   `-–node-status-update-frequency`.
+
+2. Kubernetes controller manager checks the statuses of Kubelets every
+   `–-node-monitor-period`.
+
+3. In case the status is updated  within `--node-monitor-grace-period` of time,
+   Kubernetes controller manager considers healthy status of Kubelet.
+
+> Kubernetes controller manager and Kubelets work asynchronously. It means that
+> the delay may include network latency, API Server latency, etcd latency,
+> latency caused by load on one's master nodes and so on. So if
+> `--node-status-update-frequency` is set to 5s in reality it may appear in
+> etcd in 6-7 seconds or even longer when etcd cannot commit data to quorum
+> nodes.
+
+# Failure
+
+Kubelet will try to make `nodeStatusUpdateRetry` post attempts. Currently
+`nodeStatusUpdateRetry` is constantly set to 5 in
+[kubelet.go](https://github.com/kubernetes/kubernetes/blob/release-1.5/pkg/kubelet/kubelet.go#L102).
+
+Kubelet will try to update the status in
+[tryUpdateNodeStatus](https://github.com/kubernetes/kubernetes/blob/release-1.5/pkg/kubelet/kubelet_node_status.go#L345)
+function. However, `http.Client()` timeout is not defined yet. Thus there may
+be some glitches when API Server is overloaded while TCP connection is
+established.
+
+So, there will be `nodeStatusUpdateRetry` * `--node-status-update-frequency`
+attempts to set a status of node.
+
+At the same time Kubernetes controller manager will try to check
+`nodeStatusUpdateRetry` times every `--node-monitor-period` of time. After
+`--node-monitor-grace-period` it will consider node unhealthy. It will remove
+its pods based on `--pod-eviction-timeout`
+
+> Every equation is average as
+> ```
+> node-status-update-frequency * nodeStatusUpdateRetry != node-monitor-grace-period
+> ```
+> due to latencies created by all components.
+
+Kube proxy has a watcher over API. Once pods are evicted the proxy will notice
+and will update iptables of the node. It will remove endpoints from services
+so pods from failed node won't be accessible anymore.
+
+# Recommendations for different cases
+
+## Fast Update and Fast Reaction
+
+If `-–node-status-update-frequency` is set to **4s** (10s is default).
+`--node-monitor-period` to **2s** (5s is default).
+`--node-monitor-grace-period` to **20s** (40s is default).
+`--pod-eviction-timeout` is set to **30s** (5m is default)
+
+In such scenario, pods will be evicted in **50s** as node will be considered as
```

Comment:
**50s** because the node will be considered as down after 20s, and `--pod-eviction-timeout` occurs after 30s more.
^ Suggestion if you like.